### ISSUE TITLE
`@code_warntype` is not smart in the presence of kwargs

### ISSUE THREAD

**Issue by oscardssmith** [JuliaLang/julia#45297]
For simple cases, this works due to inlining
```
julia> f(a=2) = exp(a);

julia> @code_warntype f(3)
MethodInstance for f(::Int64)
  from f(a) in Main at REPL[20]:1
Arguments
  #self#::Core.Const(f)
  a::Int64
Body::Float64
1 ─ %1 = Main.exp(a)::Float64
└──      return %1
```
but for functions where the body isn't inlined, all `@code_warntype` shows is the lowering of the keyword arguments.
```
julia> function g(x, y=3)
           z =23
           for i in 1:100
               res =1
               for j = 1:100
                   res += x
               end
               z -= res+y
           end
           z
       end
g (generic function with 2 methods)

julia> @code_warntype g(3)
MethodInstance for g(::Int64)
  from g(x) in Main at REPL[18]:1
Arguments
  #self#::Core.Const(g)
  x::Int64
Body::Int64
1 ─ %1 = (#self#)(x, 3)::Int64
└──      return %1
```
Similar problems also exist for `@code_typed` and the rest, but this (IMO) matters more for `@code_warntype` since it is (or at least should be) much more broadly used.

**Comment by oscardssmith** [comment]
Good point. Simplified. That said, I still think `@code_warntype` should be smarter here.

**Comment by simeonschaub** [comment]
What do you mean by smarter though? I'm not even sure it could just and give you the method with all optional arguments specified without resorting to somewhat arbitrary heuristics that actually go through the source code. You can ask for the optimized code if you want to see the results after inlining.

**Comment by KristofferC** [comment]
I don't really get what there is to do here. Should it just be closed? Otherwise, perhaps some elaboration on what the actual problem is could be useful.

**Comment by kockahonza** [comment]
Hi, I've come across this thread, but I have to say I agree that this is a pretty major limitation of `@code_warntype` at least as far as I can tell. While this thread talks about optional arguments, pretty much the same things applies to keyword arguments too where if you have them then the macro just shows how it unpacks/deals with the arguments and that it then calls some internal function generated from the code (at least as far as I can tell). In other words it cannot help with debugging typing issues whenever there are any optional or keyword arguments, significantly restricting its usage. Or perhaps this is just an issue with me not knowing how to do things, but is there any way to delve into the type stability of what happens in such a function using just the macros in the language itself? (I mean not using [Cthulhu](https://github.com/JuliaDebug/Cthulhu.jl)). Pretty much this same thing was also discussed [here](https://discourse.julialang.org/t/use-of-code-warntype-on-a-function-with-keyword-arguments/5521).
